Add iframe bridge documentation (#2810)
vivid-planet/comet: docs/docs/preview/iframebridge/IFrameBridge.md

---
title: IFrameBridge
sidebar_position: 8
---

The Block Preview allows you to preview your blocks in the context of the Comet Admin. The Block Preview is an iframe that is embedded in the Comet Admin. Comet Admin is intended to be a headless CMS, but to offer the capability to render a preview of the content, the block-preview got introduced.

![IFrame](images/iframe.png)

> The `IFrameBridge` is a communication bridge between the Comet Admin and the Block Preview implemented in the site's client. It allows you to send messages from the Comet Admin to the Block Preview and vice versa. In the admin client the `IFrameBridge` is already implemented and can be used out of the box and is already sending messages to the iframe site client's block preview. In the site client you can use the IFrameBridge from `@comet/cms-site` to receive/send messages from/to the Comet Admin.

Reviewer: Regarding "site's client": Isn't site and client the same?

Author: agree, i though it would be a little bit clearer when haveing both words, but we can reduce it to client.

## IFrameMessages

There are several messages that can be sent between the Comet Admin and the Block Preview. All of them are handled or are available in a higher level api which is provided in the `useIFrameBridge` hook.

> | Property                       | Sender | Description                                                                                                                                                                                                   | available                                  |

Reviewer: Out of scope: Maybe this table could be generated from source somehow.

| `IFrameHoverComponentMessage`  | Site   | The `IFrameHoverComponentMessage` is sent from the site to the admin, when a component is hovered in the iFrame and should be highlighted in the admin interface.                                             | `iFrameBridge.sendHoverComponent("route")` |
| `IFrameOpenLinkMessage`        | Site   | @deprecated                                                                                                                                                                                                   |
| `IFrameLocationMessage`        | Site   | @deprecated                                                                                                                                                                                                   |
| `IAdminBlockMessage`           | Admin  | The `IAdminBlockMessage` is sent from the admin to the site, which block should be displayed in the iFrame.                                                                                                   | `iFrameBridge.sendBlockState()`            |

> | `IAdminShowOnlyVisibleMessage` | Admin  | The `IAdminBlockMessage` is sent from the admin to the site, which block should be displayed in the iFrame.                                                                                                   | `iFrameBridge.sendShowOnlyVisible(true)`   |

Reviewer 2: This looks like a copy mistake:

```suggestion
| `IAdminShowOnlyVisibleMessage` | Admin  | The `IAdminShowOnlyVisibleMessage` is sent from the admin to the site, indicating if hidden blocks should be rendered or not.                                                                                                   | `iFrameBridge.sendShowOnlyVisible(true)`   |
```

| `IAdminHoverComponentMessage`  | Admin  | The `IAdminHoverComponentMessage` is sent from the admin to the site, when a component is hovered in the admin interface.                                                                                     | `iFrameBridge.sendHoverComponent("id")`    |
| `IAdminContentScopeMessage`    | Admin  | The `IAdminContentScopeMessage` is sent from the admin to the site, when the content scope is changed in the admin interface.                                                                                 | `iFrameBridge.sendContentScope(scope)`     |

## block-preview

To make the Block Preview embeddable in the Comet Admin, the client must provide an entry point where the client application renders the block preview and implements the `IFrameBridge` for communication. In a typical comet application the block preview is available at `site-domain.com/block-preview/page`

> To Implement the `IFrameBridge` in any site client, you can use `IFrameBridgeProvider` and `useIFrameBridge` from `@comet/cms-site`.

Reviewer: I believe you'd use the higher level `BlockPreviewProvider`

Author: I was not aware of the BlockPreviewProvider, but i don't think that the BlockPreviewProvider is here the way to go. 

The only thing BlockPreviewProvider is doing, is storing a state of the previewType and if Preview Skeletons should be shown.

<img width="1008" alt="Screenshot 2024-11-29 at 08 50 00" src="https://github.com/user-attachments/assets/a5dac0c4-2ec0-4d82-908a-896197eac416">

If you want to communicate between admin and site, one has to use the IFrameBridge.
